sustainml_cpp/include/sustainml_cpp/core/RequestReplier.hpp

```diff
@@ -0,0 +1,142 @@
+#include <fastdds/dds/publisher/DataWriter.hpp>
+#include <fastdds/dds/publisher/Publisher.hpp>
+#include <fastdds/dds/publisher/qos/DataWriterQos.hpp>
+#include <fastdds/dds/publisher/qos/PublisherQos.hpp>
+#include <fastdds/dds/subscriber/DataReader.hpp>
+#include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
+#include <fastdds/dds/subscriber/qos/SubscriberQos.hpp>
+#include <fastdds/dds/subscriber/Subscriber.hpp>
+
+namespace sustainml {
+namespace core {
+
+/**
+ * @brief This class in charge of sending request or response
+ * and listen for "response" or "request" respectively.
+ *
+ */
+class RequestReplier
+{
+
+    friend class Node;
+    friend class OrchestratorNode;
+
+public:
+
+    RequestReplier(
+            std::function<void(void*)> callback,
+            const char* topicw,
+            const char* topicr);
+
+    ~RequestReplier();
+
+    /**
+     * @brief Method used to send the response to the request in the Nodes.
+     *
+     * @param res Response message
+     */
+    void write_res(
+            ResponseTypeImpl* res);
+
+    /**
+    * @brief Method used to send the request of configuration in the OrchestratorNode.
+     *
+     * @param req Request message
+     */
+    void write_req(
+            RequestTypeImpl* req);
+
+protected:
+
+    std::function<void(void*)> callback_;
+    const char* topicr_;
+    const char* topicw_;
+
+    eprosima::fastdds::dds::DomainParticipant* participant_;
+
+    eprosima::fastdds::dds::Publisher* publisher_;
+
+    eprosima::fastdds::dds::Subscriber* subscriber_;
+
+    eprosima::fastdds::dds::TypeSupport typeRes_;
+    eprosima::fastdds::dds::TypeSupport typeReq_;
+
+    eprosima::fastdds::dds::Topic* topicR_;
+    eprosima::fastdds::dds::Topic* topicW_;
+
+    eprosima::fastdds::dds::DataWriter* writer_;
+
+    eprosima::fastdds::dds::DataReader* reader_;
+
+private:
+
+    class RequestReplyControlListener : public eprosima::fastdds::dds::DataReaderListener
+    {
+    public:
+
+        RequestReplyControlListener(
+                RequestReplier* node);
+
+        virtual ~RequestReplyControlListener();
+
+        /**
+         * @brief Callback executed when a new sample is available on the DataReader.
+         *
+         * @param reader The DataReader having new available samples.
+         */
+        void on_data_available(
+                eprosima::fastdds::dds::DataReader* reader);
+
+        /**
+         * @brief Callback executed when a DataReader matching status change.
+         *
+         * @param reader The DataReader.
+         * @param status The status of the subscription.
+         */
+        void on_subscription_matched(
+                eprosima::fastdds::dds::DataReader* reader,
+                const eprosima::fastdds::dds::SubscriptionMatchedStatus& status);
+
+    private:
+
+        RequestReplier* node_;
+        size_t matched_;
+        RequestTypeImpl req_;
+        ResponseTypeImpl res_;
```

Comment:
We should change these to what we finally agreed:

* Here, remove `RequestTypeImpl req_;`,`ResponseTypeImpl res_;` and instead have a member `void* data_` in `RequestReplier`.
* Add a new parameter to the constructor of `RequestReplier`, `void* data`
* Add a member `RequestTypeImpl req_data_;` in `NodeImpl`
* Add the req_data_ address to `NodeImpl`'s constructor `req_res_ = new RequestReplier(..., &req_data_);`
* Remove the `if-else` in the `on_data_available()` in this class (I will refer to it in the review).